File: cpp/tests/systemintegration-tests/LocalDiscoveryTest.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include <gtest/gtest.h>
#include <gmock/gmock.h>

#include "joynr/MessagingSettings.h"
#include "joynr/Settings.h"
#include "joynr/tests/DefaulttestProvider.h"
#include "joynr/tests/testProxy.h"
#include "joynr/types/ProviderQos.h"

#include "joynr/JoynrClusterControllerRuntime.h"
#include "tests/utils/PtrUtils.h"
#include "tests/JoynrTest.h"

using namespace ::testing;
using namespace joynr;

namespace joynr
{

class LocalDiscoveryTest : public ::testing::Test
{
public:
    LocalDiscoveryTest() : runtime1(), runtime2(), testDomain("testDomain")
    {
        auto settings1 =
                std::make_unique<Settings>("test-resources/MqttSystemIntegrationTest1.settings");
        auto settings2 =
                std::make_unique<Settings>("test-resources/MqttSystemIntegrationTest2.settings");

        MessagingSettings messagingSettings1(*settings1);
        MessagingSettings messagingSettings2(*settings2);

        messagingSettings1.setMessagingPropertiesPersistenceFilename(
                "End2EndBroadcastTest-runtime1-joynr.settings");
        messagingSettings2.setMessagingPropertiesPersistenceFilename(
                "End2EndBroadcastTest-runtime2-joynr.settings");

        Settings integration1Settings{"test-resources/libjoynrSystemIntegration1.settings"};
        Settings::merge(integration1Settings, *settings1, false);

        runtime1 = std::make_shared<JoynrClusterControllerRuntime>(
                std::move(settings1), failOnFatalRuntimeError);
        runtime1->init();

        Settings integration2Settings{"test-resources/libjoynrSystemIntegration2.settings"};
        Settings::merge(integration2Settings, *settings2, false);

        runtime2 = std::make_shared<JoynrClusterControllerRuntime>(
                std::move(settings2), failOnFatalRuntimeError);
        runtime2->init();

        runtime1->start();
        runtime2->start();

        discoveryQos.setArbitrationStrategy(DiscoveryQos::ArbitrationStrategy::LAST_SEEN);
        discoveryQos.setDiscoveryTimeoutMs(3000);
        discoveryQos.setRetryIntervalMs(250);

        messagingQos.setTtl(500);
    }

    ~LocalDiscoveryTest() override
    {

        runtime1->shutdown();
        test::util::resetAndWaitUntilDestroyed(runtime1);

        runtime2->shutdown();
        test::util::resetAndWaitUntilDestroyed(runtime2);

        test::util::removeAllCreatedSettingsAndPersistencyFiles();
    }

protected:
    void registerProvider(std::shared_ptr<JoynrClusterControllerRuntime> runtime)
    {
        auto testProvider = std::make_shared<tests::DefaulttestProvider>();
        joynr::types::ProviderQos providerQos;
        runtime->registerProvider<tests::testProvider>(testDomain, testProvider, providerQos);
    }

    std::shared_ptr<JoynrClusterControllerRuntime> runtime1;
    std::shared_ptr<JoynrClusterControllerRuntime> runtime2;
    const std::string testDomain;
    DiscoveryQos discoveryQos;
    MessagingQos messagingQos;

private:
    DISALLOW_COPY_AND_ASSIGN(LocalDiscoveryTest);
};

} // namespace joynr

class LocalDiscoveryTestTestProxy : public tests::testProxy
{
public:
    LocalDiscoveryTestTestProxy(
            std::weak_ptr<joynr::JoynrRuntimeImpl> runtime,
            std::shared_ptr<joynr::JoynrMessagingConnectorFactory> connectorFactory,
            const std::string& domain,
            const joynr::MessagingQos& qosSettings)
            : joynr::ProxyBase(runtime, connectorFactory, domain, qosSettings),
              testProxyBase(runtime, connectorFactory, domain, qosSettings),
              testFireAndForgetProxy(runtime, connectorFactory, domain, qosSettings),
              testSyncProxy(runtime, connectorFactory, domain, qosSettings),
              testAsyncProxy(runtime, connectorFactory, domain, qosSettings),
              testProxy(runtime, connectorFactory, domain, qosSettings)
    {
    }

    using tests::testProxy::_providerDiscoveryEntry;
};

TEST_F(LocalDiscoveryTest, testLocalLookup)
{
    registerProvider(runtime1);

    std::shared_ptr<ProxyBuilder<LocalDiscoveryTestTestProxy>> testProxyBuilder(
            runtime1->createProxyBuilder<LocalDiscoveryTestTestProxy>(testDomain));

    std::shared_ptr<LocalDiscoveryTestTestProxy> testProxy(
            testProxyBuilder->setMessagingQos(messagingQos)
                    ->setDiscoveryQos(discoveryQos)
                    ->build());

    EXPECT_TRUE(testProxy->_providerDiscoveryEntry.getIsLocal());
}

TEST_F(LocalDiscoveryTest, testGloballLookup)
{
    registerProvider(runtime1);

    std::shared_ptr<ProxyBuilder<LocalDiscoveryTestTestProxy>> testProxyBuilder(
            runtime2->createProxyBuilder<LocalDiscoveryTestTestProxy>(testDomain));

    std::shared_ptr<LocalDiscoveryTestTestProxy> testProxy(
            testProxyBuilder->setMessagingQos(messagingQos)
                    ->setDiscoveryQos(discoveryQos)
                    ->build());

    EXPECT_FALSE(testProxy->_providerDiscoveryEntry.getIsLocal());
}

TEST_F(LocalDiscoveryTest, testAsyncRegistration)
{
    auto testProvider = std::make_shared<tests::DefaulttestProvider>();
    joynr::types::ProviderQos providerQos;
    auto onSuccess = []() {};
    auto onError = [](const exceptions::JoynrRuntimeException&) { FAIL(); };

    runtime1->registerProviderAsync<tests::testProvider>(
            testDomain, testProvider, providerQos, onSuccess, onError);

    std::shared_ptr<ProxyBuilder<LocalDiscoveryTestTestProxy>> testProxyBuilder(
            runtime2->createProxyBuilder<LocalDiscoveryTestTestProxy>(testDomain));

    std::shared_ptr<LocalDiscoveryTestTestProxy> testProxy(
            testProxyBuilder->setMessagingQos(messagingQos)
                    ->setDiscoveryQos(discoveryQos)
                    ->build());
}

File: javascript/libjoynr-js/src/test/js/node_integration/consumer/LocalDiscoveryTest.js
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

let joynr = require("../../../../main/js/joynr");
const RadioProxy = require("../../../generated/joynr/vehicle/RadioProxy");
const TestWithVersionProvider = require("../../../generated/joynr/tests/TestWithVersionProvider");
const TestWithVersionProxy = require("../../../generated/joynr/tests/TestWithVersionProxy");
const IntegrationUtils = require("../IntegrationUtils");
const provisioning = require("../../../resources/joynr/provisioning/provisioning_cc");
const DiscoveryQos = require("../../../../main/js/joynr/proxy/DiscoveryQos");
const JoynrException = require("../../../../main/js/joynr/exceptions/JoynrException");

describe("libjoynr-js.integration.localDiscoveryTest", () => {
    let provisioningSuffix;
    let domain;
    let childId;
    const MyTestWithVersionProvider = function() {};

    afterEach(done => {
        IntegrationUtils.shutdownLibjoynr()
            .then(() => {
                done();
                return null;
            })
            .catch(() => {
                throw new Error("shutdown ChildProcess and Libjoynr failed");
            });
    });

    beforeEach(done => {
        provisioningSuffix = `LocalDiscoveryTest-${Date.now()}`;
        domain = provisioningSuffix;

        provisioning.channelId = provisioningSuffix;
        joynr.loaded = false;
        joynr.selectRuntime("inprocess");

        joynr
            .load(provisioning)
            .then(newjoynr => {
                joynr = newjoynr;
                IntegrationUtils.initialize(joynr);
                done();
                return null;
            })
            .catch(error => {
                if (error instanceof JoynrException) {
                    done.fail(`error in beforeEach: ${error.detailMessage}`);
                } else {
                    done.fail(`error in beforeEach: ${error}`);
                }
            });
    });

    function registerGlobalDiscoveryEntry() {
        return IntegrationUtils.initializeChildProcess(
            "TestEnd2EndCommProviderProcess",
            provisioningSuffix,
            domain
        ).then(newChildId => {
            childId = newChildId;
            return IntegrationUtils.startChildProcess(childId);
        });
    }

    function unregisterGlobalDiscoveryEntry() {
        return IntegrationUtils.shutdownChildProcess(childId);
    }

    function buildProxyForGlobalDiscoveryEntry() {
        return joynr.proxyBuilder.build(RadioProxy, {
            domain,
            messagingQos: new joynr.messaging.MessagingQos(),
            discoveryQos: new DiscoveryQos()
        });
    }

    it("local discovery entry is forwarded to proxy", done => {
        const providerQos = new joynr.types.ProviderQos({
            customParameters: [],
            priority: Date.now(),
            scope: joynr.types.ProviderScope.GLOBAL,
            supportsOnChangeSubscriptions: true
        });

        const testWithVersionProviderImpl = new MyTestWithVersionProvider();
        const testWithVersionProvider = joynr.providerBuilder.build(
            TestWithVersionProvider,
            testWithVersionProviderImpl
        );
        joynr.registration
            .registerProvider(domain, testWithVersionProvider, providerQos)
            .then(() => {
                return joynr.proxyBuilder.build(TestWithVersionProxy, {
                    domain,
                    messagingQos: new joynr.messaging.MessagingQos(),
                    discoveryQos: new DiscoveryQos()
                });
            })
            .then(testWithVersionProxy => {
                expect(testWithVersionProxy.providerDiscoveryEntry.isLocal).toBeDefined();
                expect(testWithVersionProxy.providerDiscoveryEntry.isLocal).toBe(true);
                joynr.registration.unregisterProvider(domain, testWithVersionProvider).then(() => {
                    done();
                    return null;
                });
            });
    });

    it("global discovery entry is forwarded to proxy", done => {
        registerGlobalDiscoveryEntry()
            .then(() => {
                return buildProxyForGlobalDiscoveryEntry();
            })
            .then(radioProxy => {
                expect(radioProxy.providerDiscoveryEntry.isLocal).toBeDefined();
                expect(radioProxy.providerDiscoveryEntry.isLocal).toBe(false);
                return unregisterGlobalDiscoveryEntry();
            })
            .catch(e => {
                return unregisterGlobalDiscoveryEntry()
                    .then(() => {
                        throw e;
                    })
                    .catch(e2 => {
                        throw new Error(`Error1: ${e}\n Error2: ${e2}`);
                    });
            })
            .then(() => {
                done();
                return null;
            });
    });

    it("cached global discovery entry is forwarded to proxy", done => {
        registerGlobalDiscoveryEntry()
            .then(() => {
                // build proxy to fill global capabilities cache
                return buildProxyForGlobalDiscoveryEntry();
            })
            .then(() => {
                // build proxy from global capabilities cache
                return buildProxyForGlobalDiscoveryEntry();
            })
            .then(radioProxy => {
                expect(radioProxy.providerDiscoveryEntry.isLocal).toBeDefined();
                expect(radioProxy.providerDiscoveryEntry.isLocal).toBe(false);
                return unregisterGlobalDiscoveryEntry();
            })
            .catch(e => {
                return unregisterGlobalDiscoveryEntry()
                    .then(() => {
                        throw e;
                    })
                    .catch(e2 => {
                        throw new Error(`Error1: ${e}\n Error2: ${e2}`);
                    });
            })
            .then(() => {
                done();
                return null;
            });
    });
}); // describe
